lib/class-wp-theme-json-schema-v0.php

```diff
@@ -0,0 +1,339 @@
+<?php
+/**
+ * Processes structures that adhere to the theme.json schema V0.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Class that encapsulates the processing of
+ * structures that adhere to the theme.json V0.
+ */
+class WP_Theme_JSON_Schema_V0 {
+
+	/**
+	 * How to address all the blocks
+	 * in the theme.json file.
+	 */
+	const ALL_BLOCKS_NAME = 'defaults';
+
+	/**
+	 * How to address the root block
+	 * in the theme.json file.
+	 *
+	 * @var string
+	 */
+	const ROOT_BLOCK_NAME = 'root';
+
+	/**
+	 * Data schema of each block within a theme.json.
+	 *
+	 * Example:
+	 *
+	 * {
+	 *   'block-one': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'background': 'color'
+	 *       }
+	 *     },
+	 *     'settings': {
+	 *       'color': {
+	 *         'custom': true
+	 *       }
+	 *     }
+	 *   },
+	 *   'block-two': {
+	 *     'styles': {
+	 *       'color': {
+	 *         'link': 'color'
+	 *       }
+	 *     }
+	 *   }
+	 * }
+	 */
+	const SCHEMA = array(
+		'customTemplates' => null,
+		'templateParts'   => null,
+		'styles'          => array(
+			'border'     => array(
+				'radius' => null,
+				'color'  => null,
+				'style'  => null,
+				'width'  => null,
+			),
+			'color'      => array(
+				'background' => null,
+				'gradient'   => null,
+				'link'       => null,
+				'text'       => null,
+			),
+			'spacing'    => array(
+				'padding' => array(
+					'top'    => null,
+					'right'  => null,
+					'bottom' => null,
+					'left'   => null,
+				),
+			),
+			'typography' => array(
+				'fontFamily'     => null,
+				'fontSize'       => null,
+				'fontStyle'      => null,
+				'fontWeight'     => null,
+				'lineHeight'     => null,
+				'textDecoration' => null,
+				'textTransform'  => null,
+			),
+		),
+		'settings'        => array(
+			'border'     => array(
+				'customRadius' => null,
+				'customColor'  => null,
+				'customStyle'  => null,
+				'customWidth'  => null,
+			),
+			'color'      => array(
+				'custom'         => null,
+				'customGradient' => null,
+				'gradients'      => null,
+				'link'           => null,
+				'palette'        => null,
+			),
+			'spacing'    => array(
+				'customPadding' => null,
+				'units'         => null,
+			),
+			'typography' => array(
+				'customFontSize'        => null,
+				'customLineHeight'      => null,
+				'dropCap'               => null,
+				'fontFamilies'          => null,
+				'fontSizes'             => null,
+				'customFontStyle'       => null,
+				'customFontWeight'      => null,
+				'customTextDecorations' => null,
+				'customTextTransforms'  => null,
+			),
+			'custom'     => null,
+			'layout'     => null,
+		),
+	);
+
+	/**
+	 * Utility to convert v0 schemas into v1.
+	 *
+	 * @param array $old A theme.json in v0 format.
+	 *
+	 * @return array The new theme.json in the v1 format.
+	 */
+	public static function to_v1( $old ) {
```

Comment:
Should this method be named something like "toLatest" or  "toNext" depending on whether we think the migration will be done on one step by the framework or that it goes through all the previous versions?

Reply:
I actually wonder whether it should be more just something like `parse` and the idea is that all versions including the latest should implement it (from raw config to the parsed one understood by the WP framework).

in the latest version it could be as simple as just returning the array itself.

Reply:
I went back and forth between the two options (have all schemas migrate to the latest version or a particular schema only migrates to the next). Here are the trade-offs I considered:

**Data**. In this migration, we are removing the distinction between `defaults` and `root` which had some advantages https://github.com/WordPress/gutenberg/pull/28533#issuecomment-768945224 and it's now merged into one. What if v2 wanted to introduce back the idea of "defaults"?
- With schemas that migrate to the last version: we can do it as the v0 schema still have access to defaults.
- With schemas that migrate to the next version: we can't because v1 no longer has the defaults.

**Introducing changes**. How do we introduce changes from the plugin when this is in core?
- With schemas that migrate to the last version: every schema has to have a hook and the plugin needs to hook into all of them to change how they operate (in addition to the hooks necessary for the operations themselves: merge, get_stylesheet, etc).
- With schemas that migrate to the next version: we only need one hook (in the schema loader) that allows migrating the latest version to the desired changes.

An approach in which schemas migrate to the next version seems simpler. I'm not sure the case I've shared for data is one we'd encounter. That's my rationale to opt for this approach but, definitely, the alternative is viable as well (and if we go that route, yes, what you mentioned at https://github.com/WordPress/gutenberg/pull/30600#discussion_r609449550 and https://github.com/WordPress/gutenberg/pull/30600#pullrequestreview-631124965 makes sense to me).

Reply:
I have a preference for 1 (last version) personally. I agree it requires us to write a bit more code when adding new versions but we can actually call the next "parse" inside the previous one which mean we have the flexibility of both: chaining or not.

 - if for instance v2 need to restore defaults, v0 can just return its values without calling v1
 - if v2 is just a new set of transformations, v0 can transform to v1 and then explicitly call the v1 parse function itself.

The other advantage is that we can do optimizations as well and avoid looping multiple times if something is closetful.

Reply:
:+1: Let's do that.